FDHA-6841 [E2E][Web Patient] Onboarding Practice

alaninfo23 opened the pull request:
## Description 📝

The patient should be able to respond to the Onboarding Practice Survey and the system must show an error if the patient does not enter the required fields.

Obs: I also refactored some existing functions.

## Demo

https://github.com/alaninfo23/e2e-web-patient/assets/17017218/287471dd-45d3-4ef8-aa63-142eb23250d7

## Ticket 🎟️

[FDHA-6841](https://faeththerapeutics.atlassian.net/browse/FDHA-6841)

alaninfo23 committed:
FDHA-6841 [E2E][Web Patient] Onboarding Practice

norikoStefany-daitan commented on cypress/e2e/helpers/adminHelper/adminHelper.ts:
```diff
@@ -47,7 +47,7 @@ export const loginWebAdmin = (email: string, senha: string) => {
 export const getPreviousHour = () => {
   const currentTime = moment();
-  const newTime = currentTime.clone().subtract(1, "hour");
+  const newTime = currentTime.clone(); //.subtract(10, "minutes");
```
Could we remove the comment?

alaninfo23 replied:
Thanks for letting me know, I commented just to test and forgot to remove.

norikoStefany-daitan replied:
Thanks! We can leave the `.subtract(1, "hour");`  as function name is `getPreviousHour`

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -46,14 +55,14 @@ export const SCHEDULE_TRIWEEKLY = (day: string, time: string) =>
-  schedule: string,
-  status: string,
-) => {
+export const OPTION_RADIO_EXCELLENT = 'input[value="Excellent"]';
```
Just a suggestion =)
```suggestion
export const RADIO_OPTION_EXCELLENT = 'input[value="Excellent"]';
```

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -46,14 +55,14 @@ export const SCHEDULE_TRIWEEKLY = (day: string, time: string) =>
+export const OPTION_RADIO_GOOD = 'input[value="Good"]';
+export const OPTION_RADIO_CONFUSING = 'input[value="Confusing"]';
+
+export const currentDate = moment().format("MMM DD");
```
Hmm.. since this is a variable the gets the current time it doesn't makes sense to add it on a helper file.
Probably this current time that you want is the current time when running the test, so ideally it should be on the test, or right before you use it.

alaninfo23 replied:
Ok thanks, I removed it and put it in the test, it's because I thought that if I did it this way it would reduce the repetition of this variable in each test, what do you think?

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -82,6 +91,117 @@ export const verifyWeightScreenContent = (
   );
 };
 
+export const verifyOnboardingExperienceOptions = (
```
Just to keep same pattern as the others
```suggestion
export const verifyOnboardingExperienceScreen = (
```

alaninfo23 replied:
It's true, it gets better.

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -82,6 +91,117 @@ export const verifyWeightScreenContent = (
+  cy.contains(HOW_HAS_YOUR_ONBOARDING_EXPERIENCE_BEEN_SO_FAR);
+  cy.contains(REQUIRED);
+
+  cy.get(PROGRESS_BAR_VALUE_ID(progressBarValue)).should("exist");
+  cy.get(VALUE_PROGRESS_BAR_TEXT).should(
+    "contain",
+    PERCENT_NUMBER(percentValue),
+  );
```
We could try to add these steps in the screen order, just to make it easier for the reviewer, so the progress bar validation would be under the close button validation

alaninfo23 replied:
Perfect Stef, thanks for the tip. So I do the same thing in the other functions too, ok?

norikoStefany-daitan replied:
Perfect, yes pls apply to all as well =)

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -82,6 +91,117 @@ export const verifyWeightScreenContent = (
+  cy.contains("button", NEXT).should("be.visible");
+};
+
+export const verifyPleaseProvideFeedbackScreen = (
```
Also, just suggestions =)
```suggestion
export const verifyProvideFeedbackScreen = (
```

or 

```suggestion
export const verifyFeedbackScreen = (
```

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -82,6 +91,117 @@ export const verifyWeightScreenContent = (
+  const formattedDateText = format(currentDate, "EEEE, MMMM dd");
+  cy.get(DATE_TEXT).should("contain", formattedDateText);
+
+  cy.contains(HOW_HAS_YOUR_ONBOARDING_EXPERIENCE_BEEN_SO_FAR);
```
hmm interesting, can we use contains without an assertion?

alaninfo23 replied:
I didn't understand? what does it refer to?

norikoStefany-daitan replied:
I mean, we are using `contains` without making an assertion like `should("be.visible")`

norikoStefany-daitan replied:
There are some places still missing the `should("be.visible")`

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -82,6 +91,117 @@ export const verifyWeightScreenContent = (
+  cy.contains("button", NEXT).should("be.visible");
+};
+
+export const verifyQuestionsNotBeenAnswered = (
```
```suggestion
export const verifyQuestionsNotBeenAnsweredScreen = (
```

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -82,6 +91,117 @@ export const verifyWeightScreenContent = (
+  const formattedDateText = format(currentDate, "EEEE, MMMM dd");
+  cy.get(DATE_TEXT).should("contain", formattedDateText);
+
+  cy.contains(IF_YOU_HAVE_ANY_QUESTIONS_NOT_BEEN_ANSWERED);
```
we could add the assertion here that "Required" label should not be visible

alaninfo23 replied:
Excellent tip Stefany, I'll add it and start thinking more about these possibilities in the next tests.

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -82,6 +91,117 @@ export const verifyWeightScreenContent = (
+  cy.contains("button", NEXT).should("be.visible");
+};
+
+export const verifyOnboardPracticeConfirmation = (
```
```suggestion
export const verifyOnboardPracticeConfirmationScreen = (
```

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -82,6 +91,117 @@ export const verifyWeightScreenContent = (
+  const formattedDateText = format(currentDate, "EEEE, MMMM dd");
+  cy.get(DATE_TEXT).should("contain", formattedDateText);
+
+  cy.contains(HOW_HAS_YOUR_ONBOARDING_EXPERIENCE_BEEN_SO_FAR);
```
We should verify the answers here as well (didn't remember if we did it for the other confirmation screens)

alaninfo23 replied:
I didn't understand?

norikoStefany-daitan replied:
Sorry haha let me explain better.
We could verify the answers inside the verify function. For it you could create an object with the patient's answer and pass it to the verify function so we don't have to add a lot of parameters

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -82,6 +91,117 @@ export const verifyWeightScreenContent = (
+};
+
+export const checkAlertMessage = (message: string) => {
+  const ALERT_NOTIFICATION: string = ".MuiAlert-message";
```
It's defined on the surveyHelper as well, couldn't we use from there?

alaninfo23 replied:
It's true, I hadn't noticed, thanks!

norikoStefany-daitan commented on cypress/e2e/helpers/surveyHelper.ts:
```diff
@@ -95,10 +215,11 @@ export const verifyInputMessageError = (
   cy.contains("button", SUBMIT_SURVEY).click();
   cy.get(SNACK_BAR_ALERT_ID).should("be.visible").contains(SURVEY_COMPLETED);
-  cy.wait(7000);
+  cy.wait(waitTimeMs);
```
Why we turned it into a parameter?

alaninfo23 replied:
At some point I had problems with the time needed for the success message that appears above the logoff button to disappear, so I set the parameter to make it easier, do you think I should remove it? or put cy.wait somewhere else?

norikoStefany-daitan replied:
Hmm ok, in this case I think it's better to keep the milliseconds inside the function cause this functions is always being called with the same value, and if at some point you find out that this time needs to be updated you would need to updated it in only one place =)

norikoStefany-daitan commented on cypress/e2e/strings/surveyStrings.ts:
```diff
@@ -1,5 +1,15 @@
+export const IF_YOU_HAVE_ANY_QUESTIONS_NOT_BEEN_ANSWERED: string =
+  "If you have any questions that have not been answered yet in the onboarding, please let us know here and we will address them when we meet with you one-on-one.";
+
+export const THIS_FIELD_CANNOT_BE_LEFT_BLANK: string =
```
This could be on generalStrings, we use as error for other places too

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/noTrialSpecific/onboardPractice.cy.ts:
```diff
@@ -0,0 +1,144 @@
+import { SURVEY_CARD_BUTTON } from "../../../helpers/homeHelper";
+import { NEXT } from "../../../helpers/generalStrings";
+
+describe("notTrialSpecific - Onboarding Practice Survey", () => {
```
```suggestion
describe("Onboarding Practice Survey", () => {
```
Obs. the folder name should be `nonTrialSpecific`

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/noTrialSpecific/onboardPractice.cy.ts:
```diff
@@ -0,0 +1,144 @@
+
+    cy.get(ONBOARDING_PRACTICE_SURVEY_ID).click();
+
+    verifyOnboardingExperienceOptions("25", "25", ONBOARDING_PRACTICE_SURVEY);
```
I just noticed that these values are always the same, would it make sense to send only one parameter?

alaninfo23 replied:
Very good tip, Stefany! I took advantage and made these changes to other functions, I didn't know that you could pass 1 parameter to more of one function.

norikoStefany-daitan replied:
Yes! Maybe we could keep `percentValue` instead of `progressBarValue`. wdyt?

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/noTrialSpecific/onboardPractice.cy.ts:
```diff
@@ -0,0 +1,144 @@
+    cy.get(ONBOARDING_PRACTICE_SURVEY_ID).click();
+    cy.get(CONTAINED_PRIMARY_BUTTON).contains(NEXT).click();
+    cy.wait(2000);
+    checkAlertMessage("Please select one option");
```
We could add it to strings
And before calling check function we could add a step to verify we are on the correct screen (Same for the feedback screen)

alaninfo23 replied:
Ok, I'm not sure I understand, but I inserted a step to check if you are on the correct page before and after error validation.

norikoStefany-daitan replied:
Right! I would suggest to add the validation of screen before triggering the error

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/noTrialSpecific/onboardPractice.cy.ts:
```diff
@@ -0,0 +1,144 @@
+    cy.get(CONTAINED_PRIMARY_BUTTON).contains(NEXT).click();
+    cy.wait(2000);
+    checkAlertMessage("Please select one option");
+    cy.get(OPTION_RADIO_EXCELLENT).click();
```
I think we don't need this final step

alaninfo23 replied:
I clicked to make the error disappear from the screen, because the error icon is above the log off button.

norikoStefany-daitan replied:
Got it.. that strange, could you please double check if the error snackbar never gets dismissed?

alaninfo23 replied:
Yes, the error snackbar never gets dismissed, only if click on the screen.

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/noTrialSpecific/onboardPractice.cy.ts:
```diff
@@ -0,0 +1,144 @@
+    verifyInputMessageError(true, THIS_FIELD_CANNOT_BE_LEFT_BLANK);
+  });
+
+  it("Patient should be able to answer Onboarding Practice Survey when patient added question, #6111, #5781, #5782, #5786, #5787, #5788", () => {
```
```suggestion
  it("Patient should be able to answer Onboarding Practice Survey when patient added question, #6111, #5781, #5782, #5784, #5786, #5788", () => {
```

norikoStefany-daitan commented on cypress/e2e/tests/structureSurvey/noTrialSpecific/onboardPractice.cy.ts:
```diff
@@ -0,0 +1,144 @@
+    submitSurvey(7000);
+  });
+
+  it("Patient should be able to answer Onboarding Practice Survey when didn't add question, #6110, #5789", () => {
```
```suggestion
  it("Patient should be able to answer Onboarding Practice Survey when didn't add question, #6110, #5787, #5789", () => {
```

alaninfo23 committed:
New changes:

- Removed unnecessary comments
- Updated variable names as suggested
- Moved 'currentDate' to tests as recommended
- Organized code structure in 'verifyOnboardingExperienceOptions'
- Improved order of steps for better readability
- Simplified function parameters

alaninfo23 committed:
New changes:

- Updated function names as suggested
- Added a step to validate that the Required label should not be visible.
- Fixed some import names

alaninfo23 committed:
Moved 'currentDate' to tests as recommended

alaninfo23 committed:
Refactor: Moved string variable to a separate file, renamed the test suite description, and added '.should('be.visible')' in several tests that were missing it.

alaninfo23 committed:
Improvements:
Added date check.
Renamed the test description and corresponding folder.
Used a single parameter instead of two identical values.
Added steps to verify correct screen before triggering errors.
Added a step to check if the on-screen error goes away.

alaninfo23 committed:
test name correction to onboardingPractice
